### PR TITLE
DEP: Switch CRS mismatch future warnings to errors

### PR DESCRIPTION
**Pull request by m-richards**
This addresses the 3 places (GeoDataFrame twice and GeoSeries) where supplying crs which conflict with the underlying data was raising a FutureWarning, and switches these over to ValueErrors. Have only changed these as they are marked with `#TODO: raise error in 0.9 or 0.10`, but there are a number of other FutureWarnings throughout the code (the first of which is something that should change in this branch too):
```bash
grep -rw . -e 'FutureWarning' --exclude *.pyc | cut -d: -f1 | uniq -c
      1 ./doc/source/docs/user_guide/projections.rst
      2 ./geopandas/array.py
      1 ./geopandas/base.py
      2 ./geopandas/geodataframe.py
      5 ./geopandas/geoseries.py
      4 ./geopandas/plotting.py
      1 ./geopandas/sindex.py
      3 ./geopandas/tests/test_array.py
      4 ./geopandas/tests/test_crs.py
      2 ./geopandas/tests/test_geom_methods.py
      4 ./geopandas/tests/test_geoseries.py
      1 ./geopandas/tests/test_sindex.py
      3 ./geopandas/tools/crs.py
      2 ./geopandas/tools/sjoin.py
      2 ./geopandas/tools/tests/test_sjoin.py
      3 ./geopandas/tools/tests/test_tools.py
      1 ./geopandas/_vectorized.py
```
Not sure if there is a desire to address some of these now as well (or at least least thing about how soon they should happen).

There is perhaps some sense in dealing with other crs related future warnings all in the same release? Going through the list very quickly, these seem to fit:
- `geodataframe.py` in both `__init__` and `crs`
```python
warnings.warn(
                "Assigning CRS to a GeoDataFrame without a geometry column is now "
                "deprecated and will not be supported in the future.",
                FutureWarning,
                stacklevel=2,
            )
```
- `tools/crs.py`
```python
"explicit_crs_from_epsg is deprecated. "
        "You can set the epsg on the GeoDataFrame (gdf) using gdf.crs=epsg",
```
```python
"epsg_from_crs is deprecated. "
        "You can get the epsg code from GeoDataFrame (gdf) "
        "using gdf.crs.to_epsg()",
```

In addition, the GeoSeries warning below has been for just over 2 years now, and seems like a pretty logical change to make.
```
"""\
    You are passing non-geometry data to the GeoSeries constructor. Currently,
    it falls back to returning a pandas Series. But in the future, we will start
    to raise a TypeError instead."""
```

### CODE REVIEW

**Review by martinfleis — APPROVED** [review]
Thanks! This is great. 

Regarding other changes, I think that anything that has a warning since geopandas 0.8 or older can be turned into an error. If you decide to go for it, can you split it to individual thematic PRs so it is easier to track these deprecations?

**Comment by m-richards** [comment]
> Regarding other changes, I think that anything that has a warning since geopandas 0.8 or older can be turned into an error. If you decide to go for it, can you split it to individual thematic PRs so it is easier to track these deprecations?

Sure, I'm happy to split them out, I might start with an issue listing all the existing future warnings and the PRs can target specific ones.

File: geopandas/geoseries.py [hunk 1]
```diff
@@ -143,17 +143,12 @@ def __new__(cls, data=None, index=None, crs=None, **kwargs):
                 data = data.copy()
             else:
                 if not data.crs == crs:
-                    warnings.warn(
+                    raise ValueError(
                         "CRS mismatch between CRS of the passed geometries "
                         "and 'crs'. Use 'GeoDataFrame.set_crs(crs, "
```
Comment (martinfleis) [inline geopandas/geoseries.py:148]:
```suggestion
                        "and 'crs'. Use 'GeoSeries.set_crs(crs, "
```

File: doc/source/docs/user_guide/projections.rst [hunk 2]
```diff
@@ -121,18 +121,12 @@ projection passed to GeoSeries or GeoDataFrame during the creation:
    - Lat[north]: Geodetic latitude (degree)
    - Lon[east]: Geodetic longitude (degree)
    ...
-   >>> GeoSeries(array, crs=3395).crs  # crs=3395 is ignored as array already has CRS
-   FutureWarning: CRS mismatch between CRS of the passed geometries and 'crs'. Use 'GeoDataFrame.set_crs(crs, allow_override=True)' to overwrite CRS or 'GeoDataFrame.to_crs(crs)' to reproject geometries. CRS mismatch will raise an error in the future versions of GeoPandas.
+   >>> GeoSeries(array, crs=4326)  # crs=4326 is okay, as it matches the existing CRS
+   >>> GeoSeries(array, crs=3395)  # crs=3395 is forbidden as array already has CRS
+   ValueError: CRS mismatch between CRS of the passed geometries and 'crs'. Use 'GeoDataFrame.set_crs(crs, allow_override=True)' to overwrite CRS or 'GeoDataFrame.to_crs(crs)' to reproject geometries.
```
Comment (martinfleis) [inline doc/source/docs/user_guide/projections.rst:126]: This should be a GeoSeries error, not GeoDataFrame.

Comment (m-richards) [reply doc/source/docs/user_guide/projections.rst:126]: Good catch.

**Comment by martinfleis** [comment]
Thanks!

**Pull request merged**